bullet-featherstone/src/KinematicsFeatures.cc

```diff
@@ -59,60 +69,55 @@ FrameData3d KinematicsFeatures::FrameDataRelativeToWorld(
-
-        if (indexOpt2.has_value())
+        if (linkInfo2->indexInModel.has_value())
         {
-          const auto index2 = *indexOpt2;
-          FrameData data;
-          data.pose = GetWorldTransformOfLink(*model, *linkInfo2);
-          const auto &link = model->body->getLink(index2);
-          data.linearVelocity = convert(
-            link.m_absFrameTotVelocity.getLinear());
-          data.angularVelocity = convert(
-            link.m_absFrameTotVelocity.getAngular());
-          return data;
+          return getNonBaseLinkFrameData(model, linkInfo2.get());
         }
       }
     }
-
-    auto collisionIt = this->collisions.find(_id.ID());
-    if (collisionIt != this->collisions.end())
+    else
     {
-      const auto &collisionInfo = collisionIt->second;
-
-      const auto linkIt2 = this->links.find(collisionInfo->link);
-      if (linkIt2 != this->links.end())
+      auto collisionIt = this->collisions.find(_id.ID());
+      if (collisionIt != this->collisions.end())
       {
-        const auto &linkInfo2 = linkIt2->second;
-        const auto indexOpt2 = linkInfo2->indexInModel;
-        model = this->ReferenceInterface<ModelInfo>(linkInfo2->model);
+        isCollision = true;
+        const auto &collisionInfo = collisionIt->second;
 
-        if (indexOpt2.has_value())
+        const auto linkIt2 = this->links.find(collisionInfo->link);
+        if (linkIt2 != this->links.end())
         {
-          const auto index2 = *indexOpt2;
-          FrameData data;
-          data.pose = GetWorldTransformOfLink(*model, *linkInfo2);
-          const auto &link = model->body->getLink(index2);
-          data.linearVelocity = convert(
-            link.m_absFrameTotVelocity.getLinear());
-          data.angularVelocity = convert(
-            link.m_absFrameTotVelocity.getAngular());
-          return data;
+          const auto &linkInfo2 = linkIt2->second;
+          model = this->ReferenceInterface<ModelInfo>(linkInfo2->model);
+          collisionPoseOffset = collisionInfo->linkToCollision;
+          if (linkInfo2->indexInModel.has_value())
+          {
+            auto data = getNonBaseLinkFrameData(model, linkInfo2.get());
+            data.pose = data.pose * collisionPoseOffset;
+            return data;
+          }
+        }
+      }
+      else
+      {
+        auto modelIt = this->models.find(_id.ID());
+        if (modelIt != this->models.end())
+        {
+          model = modelIt->second.get();
+          isModel = true;
         }
       }
     }
-
-    if (!model || model->body == nullptr)
-      model = this->FrameInterface<ModelInfo>(_id);
   }
 
   FrameData data;
-  if(model && model->body)
+  if (model && model->body)
   {
-    data.pose = convert(model->body->getBaseWorldTransform())
-      * model->baseInertiaToLinkFrame;
+    data.pose = convert(model->body->getBaseWorldTransform());
+    if (!isModel)
+      data.pose = data.pose * model->baseInertiaToLinkFrame;
+    if (isCollision)
```

Comment:
I'm not sure why we need this here. Aren't collisions handled above starting from line 81?

Reply:
This function could probably be refactored better. The logic starting in line 81 handles collisions for 'non-base' links,  i.e. if `linkInfo2->indexInModel.has_value()` is true, it's a non-base link. There is a similar check in many places of this function - if this check fails, it all comes down to this part of code (starting line 114) that computes frame data using the base link.